# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd


try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data_pivot = pd.pivot_table(data, values='debt', index='children', aggfunc={'debt' : ['sum', 'count', 'mean']})
data_pivot['mean'] = data_pivot['mean'] * 100
data_pivot

count      mean     sum
children                           
0         14091.0  7.543822  1063.0
1          4808.0  9.234609   444.0
2          2052.0  9.454191   194.0
3           330.0  8.181818    27.0
4            41.0  9.756098     4.0
5             9.0  0.000000     0.0

In [23]:
##КОД РЕВЬЮЕРА
data_pivot1 = pd.pivot_table(data, values='debt', index='children', aggfunc=['sum', 'count', 'mean'])
data_pivot1 

sum  count      mean
          debt   debt      debt
children                       
0         1063  14091  0.075438
1          444   4808  0.092346
2          194   2052  0.094542
3           27    330  0.081818
4            4     41  0.097561
5            0      9  0.000000

**Вывод:** Люди не имеющие детей лучше всего отдают кредиты, отрыв от остальных категорий около 2-х %. Доля должников тут составляет 7.5%. Несколько хуже - люди имеющие одного или двух детей. 9.2% и 9.4%, соответственно. Такой вывод наталкивает на мысли, что с каждым ребенком доля просроченных платежей возрастает на 0.2%, однако данных для такого вывода недостаточно.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data['family_status_id'].unique()


array([0, 1, 2, 3, 4])

In [25]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [26]:
family_status_pivot = pd.pivot_table(data, values='debt', index='family_status', aggfunc={'debt' : ['sum', 'count', 'mean']})
family_status_pivot['mean'] = family_status_pivot['mean'] * 100
family_status_pivot

count      mean    sum
family_status                                  
Не женат / не замужем   2796.0  9.763948  273.0
в разводе               1189.0  7.064760   84.0
вдовец / вдова           951.0  6.624606   63.0
гражданский брак        4134.0  9.313014  385.0
женат / замужем        12261.0  7.560558  927.0

**Вывод:** Наиболее дисциплинированными заемщиками являются лица, состоящие в официальном браке, доля просроченных платежей в этой категории 7.5%.   Наименее дисциплинированными не женатые/не замужние с долей должников 9.7%.
Данные, на первый взгляд, являются достоверными.
P.S. Так как в данных столбца "family_status" отсутствуют пропуски, было решено использовать названия семейного положения, вместо их кодов из колонки "family_status_id"

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
total_income_category_pivot = pd.pivot_table(data, values='debt', index='total_income_category', aggfunc={'debt' : ['sum', 'count', 'mean']})
total_income_category_pivot['mean'] = total_income_category_pivot['mean'] * 100
total_income_category_pivot

count      mean     sum
total_income_category                           
A                         25.0  8.000000     2.0
B                       5014.0  7.060231   354.0
C                      15921.0  8.498210  1353.0
D                        349.0  6.017192    21.0
E                         22.0  9.090909     2.0

**Вывод:** В связи с недостаточным для анализа объемом выборки по уровню дохода представляется возможным сравнить только категории B и C. В категории С доля просроченных займов 8.4%, в категории B - 7%. Тем не менее этих даных недостаточно для формулировки однозначного вывода о зависимости доли просроченных кредитов от располагаемого дохода.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
purpose_category_pivot = pd.pivot_table(data, values='debt', index='purpose_category', aggfunc={'debt' : ['sum', 'count', 'mean']})
purpose_category_pivot['mean'] = purpose_category_pivot['mean'] * 100
purpose_category_pivot

count      mean    sum
purpose_category                                  
операции с автомобилем     4279.0  9.347978  400.0
операции с недвижимостью  10751.0  7.255139  780.0
получение образования      3988.0  9.252758  369.0
проведение свадьбы         2313.0  7.911803  183.0

**Вывод:** Наиболее желанными клиентами оказались те, кто планирует операции с недвижимостью, доля должников тут всего 7.2%.
Наименее желанными - автолюбители, 9.3% просроченных кредитов.
Наблюдаем значительный разрыв в % между категориями "операции с автомобилем","получение образования" с одной стороны, и "операции с недвижимостью" с другой.
Категорию "проведение свадьбы" с другими категориями сравнивать некорректно из-за недостаточной выборки, она минимум в 2 раза меньше остальных.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Наибольшее кол-во пропусков встречалось в столбцах "days_employed" и "total_income". Также в столбце "days_employed" встречалось большое кол-во отрицательных значений, что очень похоже на системную ошибку. Требуется проверить источники данных, методы их обработки и хранения. Пропуски, предположительно, появились из-за нежелания некоторых заемщиков разглашать некоторые сведения о себе по каким-то личным соображениям. В указанных выше 2х столбцах содержатся количественные показатели, именно они чаще всего располагают к неточностям и ошибкам

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианное значение менее чувствительно к случайным ошибкам, чем среднее. Также оно (медианное) дает более точную картину в целом, соответственно ее можно корректнее интерпретировать

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️ Медиана не чувствительна к выборке, все верно)) 
  </div>

### Шаг 4: общий вывод.

Исходя из результатов исследования, можно сделать вывод о том, что чем больше в семье детей, тем выше доля просроченных платежей. Меньше всего просроченных кредитов, с большим отрывом (около 2-х%) у семей не имеющих детей вовсе.
Касаемо зависимости от семейного положения. Наиболее дисциплинированными, с отрывом в 2%, оказались люди, состоящие в официально зарегестрированном браке. \
Даных недостаточно для формулировки однозначного вывода о зависимости доли просроченных кредитов от располагаемого дохода. Из 5 категорий, сгруппированных по доходам, имеется возможность сравнить лишь 2, и то необходимо учесть разницу в объеме выборки, составляющую около 3-х раз (350 человек и 1350 человек). \
Наиболее желанными клиентами оказались те, кто планирует операции с недвижимостью, доля должников тут всего 7.2%. Наименее желанными - автолюбители, 9.3% просроченных кредитов. Наблюдаем значительный разрыв в % между категориями "операции с автомобилем","получение образования" с одной стороны, и "операции с недвижимостью" с другой. Категорию "проведение свадьбы" с другими категориями сравнивать некорректно из-за недостаточной выборки, она минимум в 2 раза меньше остальных. \
Для улучшения качества и увеличения кол-ва данных необходимо расследование причин появления большого кол-ва отрицательных значений в столбце "days_employed" и большого кол-ва пропусков и столбце "total_income". Также необходимо увеличить объем данных по категориям заемщиков: с кол-вом детей более 3-х, в разводе, вдовец/вдова, кат. дохода A, D, E, проведение свадьбы